tidy3d/plugins/mode/solver.py

```diff
@@ -507,18 +506,59 @@ def any_pec(eps_vec, threshold=0.9 * np.abs(pec_val)):
             mat = dnz * mat * dnz.T
             vec_init = dnz * vec_init
 
+        # Denote the original eigenvalue problem as Ax = lambda x,
+        # with left and right preconditioners, we solve for the following eigenvalue problem,
+        # L A R y = lambda LR y, where x = R y
+        precon_left = None
+        precon_right = None
+        generalized_M = None  # matrix in the generalized eigenvalue problem
         if enable_preconditioner:
-            precon = sp.diags(1 / mat.diagonal())
-            mat = mat * precon
-        else:
-            precon = None
+            if PRECONDITIONER == "Jacobi":
+                precon_right = sp.diags(1 / mat.diagonal())
+
+            elif PRECONDITIONER == "Material":
+
+                def conditional_inverted_vec(eps_vec, threshold=1):
+                    """Returns a diagonal sparse matrix whose i-th element in the diagonal
+                    is |eps_i|^-1 if |eps_i|>threshold, and |eps_i| otherwise.
+                    """
+                    abs_vec = np.abs(eps_vec)
+                    return sp.spdiags(
+                        np.where(abs_vec > threshold, 1.0 / abs_vec, abs_vec), [0], N, N
+                    )
+
+                precon_left = sp.bmat(
+                    [
+                        [conditional_inverted_vec(mu_yy), None],
+                        [None, conditional_inverted_vec(mu_xx)],
+                    ]
+                )
+                precon_right = sp.bmat(
+                    [
+                        [conditional_inverted_vec(eps_xx), None],
+                        [None, conditional_inverted_vec(eps_yy)],
+                    ]
+                )
+            generalized_M = precon_right
+            mat = mat @ precon_right
+            if precon_left is not None:
+                generalized_M = precon_left @ generalized_M
+                mat = precon_left @ mat
+
+        # Trim small values in single precision case
+        if mat_precision == "single":
+            cls.trim_small_values(mat, tol=fp_eps)
 
         if analyze_conditioning:
             aca = mat.conjugate().T * mat
             aac = mat * mat.conjugate().T
             diff = aca - aac
-            print(spl.norm(diff, ord=np.inf), spl.norm(aca, ord=np.inf), spl.norm(aac, ord=np.inf))
-            print(spl.norm(diff, ord="fro"), spl.norm(aca, ord="fro"), spl.norm(aac, ord="fro"))
+            print(
+                f"inf-norm: A*A: {spl.norm(aca, ord=np.inf)}, AA*: {spl.norm(aac, ord=np.inf)}, nonnormality: {spl.norm(diff, ord=np.inf)}, relative nonnormality: {spl.norm(diff, ord=np.inf)/spl.norm(aca, ord=np.inf)}"
+            )
+            print(
+                f"fro-norm: A*A: {spl.norm(aca, ord='fro')}, AA*: {spl.norm(aac, ord='fro')}, nonnormality: {spl.norm(diff, ord='fro')}, relative nonnormality: {spl.norm(diff, ord='fro')/spl.norm(aca, ord='fro')}"
+            )
```

Comment:
The accuracy of eigenvalue problem is usually measured by how far the matrix is away from normal matrix (normal matrix has conditioning number =1 in solving eigenvalues). But I'm not sure which metric is a good measurement. Probably the relative one makes more sense? @caseyflex

Reply:
Yeah I agree that relative norm makes more sense here.